main/Main.hs

```diff
@@ -2,180 +2,290 @@
+            dbStructure <- getDbStructure schema
+            liftIO $
+              atomicWriteIORef refDbStructure $ Just dbStructure
         case result of
           Left e -> do
-            putStrLn ("Failed to query the database. Retrying." :: Text)
+            putStrLn
+              ("Failed to query the database. Retrying." :: Text)
             hPutStrLn stderr (toS $ encodeError e)
             work
           Right _ -> do
             atomicWriteIORef refIsWorkerOn False
             putStrLn ("Connection successful" :: Text)
+    --
+    -- used to check if the provided db-uri lets
+    -- the application access the PostgreSQL database. This method is used 
+    -- the first time the connection is tested, but only to test before 
+    -- calling 'getDbStructure' inside the 'connectionWorker' method.
 
--- | Connect to pg server if it fails retry with capped exponential backoff until success
-connectingSucceeded :: P.Pool -> IO Bool
-connectingSucceeded pool =
-  retrying (capDelay 32000000 $ exponentialBackoff 1000000)
-           shouldRetry
-           (const $ P.release pool >> isConnectionSuccessful)
-  where
-    isConnectionSuccessful :: IO Bool
-    isConnectionSuccessful = do
-      testConn <- P.use pool $ H.sql "SELECT 1"
-      case testConn of
-        Left e -> hPutStrLn stderr (toS $ encodeError e) >> pure False
-        _ -> pure True
-    shouldRetry :: RetryStatus -> Bool -> IO Bool
-    shouldRetry rs isConnSucc = do
-      delay <- pure $ fromMaybe 0 (rsPreviousDelay rs) `div` 1000000
-      itShould <- pure $ not isConnSucc
-      when itShould $
-        putStrLn $ "Attempting to reconnect to the database in " <> (show delay::Text) <> " seconds..."
-      return itShould
+    -- The connection tries are capped, but if the connection times out no error is
+    -- thrown, just 'False' is returned.
+    connectingSucceeded :: IO Bool
+    connectingSucceeded =
+      retrying
+        (capDelay 32000000 $ exponentialBackoff 1000000)
+        shouldRetry
+        (const $ P.release pool >> isConnectionSuccessful)
+      where
+        isConnectionSuccessful :: IO Bool
+        isConnectionSuccessful = do
+          testConn <- P.use pool $ H.sql "SELECT 1"
+          case testConn of
+            Left e -> hPutStrLn stderr (toS $ encodeError e) >> pure False
+            _ -> pure True
+        shouldRetry :: RetryStatus -> Bool -> IO Bool
+        shouldRetry rs isConnSucc = do
+          delay <- pure $ fromMaybe 0 (rsPreviousDelay rs) `div` 1000000
+          itShould <- pure $ not isConnSucc
+          when itShould $
+            putStrLn $
+            "Attempting to reconnect to the database in " <>
+            (show delay :: Text) <>
+            " seconds..."
+          return itShould
+
+
+{-|
+  This is where everything starts.
 
+  Things that can throw errors: loadSecretFile
+                              , readOptions
+                              , when isMalformedProxyUri
```

Comment:
I'm working on a way to detect the used libpq version and this may throw another error, this kind of comment can get outdated pretty quickly and is easy to forget to update it.
I think it's better to just leave the code to express this.